src/coprocessor/codec/batch/lazy_column_vec.rs

```diff
@@ -95,50 +95,37 @@ impl LazyBatchColumnVec {
         }
     }
 
-    /// Retains the elements according to a boolean array.
-    ///
-    /// # Panics
-    ///
-    /// Panics if `retain_arr` is not long enough.
-    pub fn retain_rows_by_array(&mut self, retain_arr: &[bool]) {
-        if self.rows_len() == 0 {
-            return;
-        }
-
-        let current_rows_len = self.rows_len();
-
-        // We retain column by column to be efficient.
-        for col in &mut self.columns {
-            assert_eq!(col.len(), current_rows_len);
-            col.retain_by_array(retain_arr);
-        }
-
-        self.assert_columns_equal_length();
-    }
-
     /// Returns maximum encoded size.
-    pub fn maximum_encoded_size(&self, output_offsets: impl AsRef<[u32]>) -> Result<usize> {
+    // TODO: Move to other place.
+    pub fn maximum_encoded_size(
+        &self,
+        logical_rows: impl AsRef<[usize]>,
+        output_offsets: impl AsRef<[u32]>,
+    ) -> Result<usize> {
+        let logical_rows = logical_rows.as_ref();
         let mut size = 0;
         for offset in output_offsets.as_ref() {
-            size += self.columns[(*offset) as usize].maximum_encoded_size()?;
+            size += self.columns[(*offset) as usize].maximum_encoded_size(logical_rows)?;
         }
         Ok(size)
     }
 
     /// Encodes into binary format.
+    // TODO: Move to other place.
     pub fn encode(
         &self,
+        logical_rows: impl AsRef<[usize]>,
         output_offsets: impl AsRef<[u32]>,
         schema: impl AsRef<[FieldType]>,
         output: &mut Vec<u8>,
     ) -> Result<()> {
-        let len = self.rows_len();
         let schema = schema.as_ref();
-        for i in 0..len {
-            for offset in output_offsets.as_ref() {
+        let output_offsets = output_offsets.as_ref();
+        for idx in logical_rows.as_ref() {
+            for offset in output_offsets {
```

Comment:
I think to make `output_offsets` be the outer loop is more cache friendly. (Even though not all logical rows adjacent)

Reply:
Yes, but TiDB will then get incorrect orders.. Currently TiDB expect data ordered by row instead of by column. This should be resolved in DAG v2 since we can stipulate a column format in DAG v2.